examples/99_in_tool_selection/ex_in_tool_selection/main.rb

```diff
@@ -0,0 +1,364 @@
+        if picked_first_point? || (@hovered_entity && @valid_hovered_entity)
+          @mouse_ip.draw(view)
+          view.tooltip = @mouse_ip.tooltip if @mouse_ip.valid?
+        end
+      end
+
+      def getExtents
+        bounds = Geom::BoundingBox.new
+        bounds.add(picked_points) unless picked_points.empty?
+        bounds
+      end
+
+      private
+
+      # Check if an entity is allowed to be selected by this tool.
+      def valid_entity?(entity)
+        # In this somewhat artificial example we only allow components to be
+        # moved. In a real life example you'd typically filter entities by
+        # other criteria.
+
+        # To replicate native Move, Rotate and Scale tools, allow all
+        # DrawingElements except Axes.
+        ### !entity.is_a?(Sketchup::Axes)
+
+        # Another typical use case is a custom tool that edits some property of
+        # a custom objects, e.g. Bezier curve control points or wall corner
+        # points. For this, you'd typically check for the entity being a group
+        # and/or component, and if it has the attribute dictionary holding your
+        # custom properties.
+        ### entity.is_a?(Sketchup::Group) && entity.attribute_dictionary("TurboWall2000")
+
+        entity.is_a?(Sketchup::ComponentInstance)
+      end
+
+      # Check if the selection as a whole is valid for this tool.
+      def valid_selection?
+        selection = Sketchup.active_model.selection
+
+        selection.size == 1 && valid_entity?(selection.first)
+      end
+
+      # Try selecting the entity under the mouse cursor, if the tool allows it.
+      def try_select_entity(view, x, y)
+        view.model.selection.clear
+
+        pickhelper = view.pick_helper(x, y)
+        @hovered_entity = pickhelper.best_picked
+        return unless @hovered_entity
+
+        @valid_hovered_entity = valid_entity?(@hovered_entity)
+        return unless @valid_hovered_entity
+
+        view.model.selection.add(@hovered_entity)
+      end
+
+      def start_move(view)
+        # HACK: create a new input point with a given position rather than
+        # copying the existing one. This fixes an issue in some cases where
+        # the points appear to get linked and the @picked_first_ip position
+        # changes when you move the mouse around, even if that should only
+        # affect @mouse_ip.
+        ### @picked_first_ip.copy!(@mouse_ip)
+        @picked_first_ip = Sketchup::InputPoint.new(@mouse_ip.position)
+
+        # Wrap model changes in an operation so it can be undone in one step.
+        # Typically we disable drawing in operations to increase performance,
+        # but here we want live updates when moving the mouse.
+        view.model.start_operation("Move", false)
+
+        # Track whether we are in an operation so we can abort it if the user
+        # activates another tool.
+        @in_operation = true
+
+        # Track the mouse position between frames so we can update the position
```

Comment:
"frames"?

Reply:
I don't really know what to call it. Mouse move perhaps? But "a move" could be interpreted as being from you start moving the mouse until you stop it, not the numerous individual events.

Reply:
I think `@last_mouse_position` is sufficient. "frames" threw me off a bit here.